src/java/org/apache/cassandra/db/marshal/TypeParser.java

```diff
@@ -130,6 +138,73 @@ public AbstractType<?> parse() throws SyntaxException, ConfigurationException
             return getAbstractType(name);
     }
 
+
+    /**
+     * parse PartitionOrdering from old version of PartitionOrdering' string format 
+     * */
+    private static  AbstractType<?> defaultParsePartitionOrdering(TypeParser typeParser)
+    {
+        IPartitioner partitioner = DatabaseDescriptor.getPartitioner();
+        Iterator<String> argIterator = typeParser.getKeyValueParameters().keySet().iterator();
+        if (argIterator.hasNext()) {
+            partitioner = FBUtilities.newPartitioner(argIterator.next());
+            assert !argIterator.hasNext();
+        }
+        return partitioner.partitionOrdering();
+    }
+    
+    //the format is (partitioner:type)
+    public AbstractType<?> getPartitionerDefinedOrder()
+    {
+        if (isEOS())
+            return defaultParsePartitionOrdering(this);
+        skipBlank();
+        if (str.charAt(idx) != '(')
+            throw new IllegalStateException();
+        Pair<Boolean, AbstractType<?>>  result = null;
+        ++idx; // skipping '('
+
+        if (str.charAt(idx) == ')')
+        {
+            ++idx;
+            return  defaultParsePartitionOrdering(this);
+        }
+        skipBlank();
+        String k = readNextIdentifier();
+        skipBlank();
+        if (str.charAt(idx) == ':')
+        {
+            ++idx;
+            skipBlank();
+        }
+        else if (str.charAt(idx) != ',' && str.charAt(idx) != ')')
+        {
+            return defaultParsePartitionOrdering(this);
+        }
+        IPartitioner partitioner = FBUtilities.newPartitioner(k);
+        AbstractType<?> type = partitioner.partitionOrdering();
+        if (partitioner.partitionOrdering() instanceof PartitionerDefinedOrder)
```

Comment:
Looks like this test should be checking if a `:` was seen instead.

Reply:
@blambov 
 It may no need to do checking.
There will be two  possibilities when arrive here,1.PartitionerDefinedOrder(<partitioner>), then the final results will returned at the catch section from line 194 to line 205 .the partitioner type is get from k
2.PartitionerDefinedOrder(<partitioner>:<baseType>),when arriving here the base type can be parsed with function parse() at line 192 ,also the partitioner type is get with k.
For ByteOrderedPartitioner AND OrderPreservingPartitioner ,their partitionOrdering() will return ByteType.Instance and UTF8Type.Instance so we should only do for Random and mum3

Reply:
I modify the code , and left some comment around this place . For format "PartitionerDefinedOrder(partitioner)" I set the idx with the initIdx and  re-use the original way of parse here . But at first a partitioner.partitionOrdering() was just retrun . Finally I change to the original way of parse for I am not sure if some other corner case may exist?